// calligraph/dist/index.js
'use client';
import { jsx, jsxs } from 'react/jsx-runtime';
import { MotionConfig, AnimatePresence, motion, useIsPresent, useMotionValue, animate, useTransform } from 'motion/react';
import { useRef, useState, useEffect, useCallback } from 'react';

const isDigit = (c)=>c >= "0" && c <= "9";
function mod(n, m) {
    return (n % m + m) % m;
}
const DIGIT_DISTANCE = 8;
const animations = {
    default: {
        duration: 0.38,
        ease: [
            0.19,
            1,
            0.22,
            1
        ]
    },
    smooth: {
        type: "spring",
        duration: 0.4,
        bounce: 0
    },
    snappy: {
        type: "spring",
        duration: 0.35,
        bounce: 0.15
    },
    bouncy: {
        type: "spring",
        duration: 0.5,
        bounce: 0.3
    }
};

function computeLCS(oldStr, newStr) {
    const m = oldStr.length;
    const n = newStr.length;
    const dp = [];
    for(let i = 0; i <= m; i++){
        dp[i] = [];
        for(let j = 0; j <= n; j++){
            if (i === 0 || j === 0) {
                dp[i][j] = 0;
            } else if (oldStr[i - 1] === newStr[j - 1]) {
                dp[i][j] = dp[i - 1][j - 1] + 1;
            } else {
                dp[i][j] = Math.max(dp[i - 1][j], dp[i][j - 1]);
            }
        }
    }
    const pairs = [];
    let i = m;
    let j = n;
    while(i > 0 && j > 0){
        if (oldStr[i - 1] === newStr[j - 1]) {
            pairs.push([
                i - 1,
                j - 1
            ]);
            i--;
            j--;
        } else if (dp[i - 1][j] > dp[i][j - 1] || dp[i - 1][j] === dp[i][j - 1] && i >= j) {
            i--;
        } else {
            j--;
        }
    }
    pairs.reverse();
    return pairs;
}
function reconcileTextKeys(prevText, newText, prevKeys, nextId) {
    const matches = computeLCS(prevText, newText);
    const newKeys = new Array(newText.length).fill("");
    for (const [oldIdx, newIdx] of matches){
        newKeys[newIdx] = prevKeys[oldIdx];
    }
    let id = nextId;
    let newCount = 0;
    for(let i = 0; i < newKeys.length; i++){
        if (!newKeys[i]) {
            newKeys[i] = `c${id++}`;
            newCount++;
        }
    }
    const keptCount = newText.length - newCount;
    const removedCount = prevText.length - keptCount;
    const totalChange = newCount + removedCount;
    const maxLen = Math.max(newText.length, prevText.length);
    return {
        keys: newKeys,
        changeRatio: maxLen > 0 ? totalChange / maxLen : 1,
        nextId: id
    };
}
function reconcileDigitKeys(prevText, newText, prevKeys, nextId) {
    const toNum = (s)=>parseFloat(s.replace(/[^0-9.-]/g, "")) || 0;
    const direction = Math.sign(toNum(newText) - toNum(prevText));
    const oldChars = prevText.split("");
    const newChars = newText.split("");
    const firstDigitIndex = (arr)=>{
        const idx = arr.findIndex((c)=>isDigit(c));
        return idx === -1 ? arr.length : idx;
    };
    const newPrefixLen = firstDigitIndex(newChars);
    const oldPrefixLen = firstDigitIndex(oldChars);
    const minPrefix = Math.min(newPrefixLen, oldPrefixLen);
    const newKeys = new Array(newChars.length);
    let id = nextId;
    for(let i = 0; i < newPrefixLen; i++){
        newKeys[i] = i < minPrefix && newChars[i] === oldChars[i] ? prevKeys[i] : id++;
    }
    const oldBody = oldChars.slice(oldPrefixLen);
    const newBody = newChars.slice(newPrefixLen);
    const oldBodyKeys = prevKeys.slice(oldPrefixLen);
    const maxBodyLen = Math.max(oldBody.length, newBody.length);
    const padOld = [
        ...Array(Math.max(0, maxBodyLen - oldBody.length)).fill(""),
        ...oldBody
    ];
    const padNew = [
        ...Array(Math.max(0, maxBodyLen - newBody.length)).fill(""),
        ...newBody
    ];
    const padKeys = [
        ...Array(Math.max(0, maxBodyLen - oldBodyKeys.length)).fill(-1),
        ...oldBodyKeys
    ];
    const bodyOffset = maxBodyLen - newBody.length;
    for(let i = 0; i < newBody.length; i++){
        const pi = bodyOffset + i;
        newKeys[newPrefixLen + i] = padNew[pi] === padOld[pi] && padKeys[pi] >= 0 ? padKeys[pi] : id++;
    }
    return {
        keys: newKeys,
        direction,
        nextId: id
    };
}

function NumberRenderer({ text, Component, transition, stagger, animateInitial, onComplete, className, style, rest }) {
    const chars = text.split("");
    const nextIdRef = useRef(chars.length);
    const [prevText, setPrevText] = useState(text);
    const [digitKeys, setDigitKeys] = useState(()=>chars.map((_, i)=>i));
    const dirRef = useRef(1);
    if (text !== prevText) {
        const result = reconcileDigitKeys(prevText, text, digitKeys, nextIdRef.current);
        nextIdRef.current = result.nextId;
        dirRef.current = result.direction;
        setDigitKeys(result.keys);
        setPrevText(text);
    }
    const dir = dirRef.current;
    const prefixLen = (()=>{
        const idx = chars.findIndex((c)=>isDigit(c));
        return idx === -1 ? chars.length : idx;
    })();
    return /*#__PURE__*/ jsx(MotionConfig, {
        transition: transition,
        children: /*#__PURE__*/ jsx(Component, {
            "aria-label": text,
            style: {
                display: "inline-flex",
                position: "relative",
                ...style
            },
            className: className,
            ...rest,
            children: /*#__PURE__*/ jsx(AnimatePresence, {
                mode: "popLayout",
                initial: animateInitial,
                children: chars.map((char, i)=>{
                    const isPrefix = i < prefixLen;
                    const outerKey = isPrefix ? `pre-${i}` : `col-${chars.length - 1 - i}`;
                    const delay = i * stagger;
                    const isLast = i === chars.length - 1;
                    return /*#__PURE__*/ jsx(motion.span, {
                        layout: "position",
                        initial: isPrefix ? false : {
                            opacity: 0
                        },
                        animate: isPrefix ? undefined : {
                            opacity: 1
                        },
                        exit: isPrefix ? undefined : {
                            opacity: 0
                        },
                        style: {
                            display: "inline-block",
                            position: "relative"
                        },
                        children: isPrefix ? /*#__PURE__*/ jsx("span", {
                            style: {
                                display: "inline-block",
                                whiteSpace: "pre"
                            },
                            children: char
                        }) : /*#__PURE__*/ jsx(AnimatePresence, {
                            mode: "popLayout",
                            initial: animateInitial,
                            propagate: true,
                            children: /*#__PURE__*/ jsx(motion.span, {
                                "aria-hidden": "true",
                                initial: {
                                    y: isDigit(char) ? dir > 0 ? DIGIT_DISTANCE : -DIGIT_DISTANCE : 0,
                                    filter: "blur(2px)",
                                    scale: 0.5,
                                    opacity: 0
                                },
                                animate: {
                                    y: 0,
                                    opacity: 1,
                                    filter: "blur(0px)",
                                    scale: 1,
                                    transition: {
                                        delay
                                    }
                                },
                                exit: {
                                    y: isDigit(char) ? dir > 0 ? -DIGIT_DISTANCE : DIGIT_DISTANCE : 0,
                                    opacity: 0,
                                    filter: "blur(2px)",
                                    scale: 0.5,
                                    transition: {
                                        delay
                                    }
                                },
                                onAnimationComplete: isLast && onComplete ? onComplete : undefined,
                                style: {
                                    display: "inline-block",
                                    whiteSpace: "pre"
                                },
                                children: char
                            }, digitKeys[i])
                        })
                    }, outerKey);
                })
            })
        })
    });
}

function DigitNum({ n, current }) {
    const y = useTransform(current, (c)=>{
        let offset = mod(n - c, 10);
        if (offset > 5) offset -= 10;
        const clamped = Math.max(-1, Math.min(1, offset));
        return `${-clamped * 100}%`;
    });
    return /*#__PURE__*/ jsx(motion.span, {
        "aria-hidden": true,
        style: {
            position: "absolute",
            top: 0,
            left: "50%",
            x: "-50%",
            display: "inline-block",
            whiteSpace: "pre",
            y
        },
        children: n
    });
}
const digits = [
    0,
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    9
];
const FADE_HEIGHT = "0.25em";
const FADE_MASK = `linear-gradient(to bottom, transparent 0%, black ${FADE_HEIGHT}, black calc(100% - ${FADE_HEIGHT}), transparent 100%)`;
function SlotColumn({ digit, direction, transition, delay, animateIn }) {
    const isPresent = useIsPresent();
    const spinIn = Math.max(digit, 1);
    const startValue = animateIn ? digit - spinIn * (direction || 1) : digit;
    const current = useMotionValue(startValue);
    const cumulativeRef = useRef(digit);
    const prevDigitRef = useRef(digit);
    const initialRef = useRef(true);
    if (digit !== prevDigitRef.current) {
        const old = prevDigitRef.current;
        let diff;
        if (direction > 0) {
            diff = digit >= old ? digit - old : 10 - old + digit;
        } else if (direction < 0) {
            diff = old >= digit ? -(old - digit) : -(10 - digit + old);
        } else {
            diff = digit - old;
        }
        cumulativeRef.current += diff;
        prevDigitRef.current = digit;
    }
    useEffect(()=>{
        if (!isPresent) {
            const spinOut = Math.max(digit, 1);
            animate(current, cumulativeRef.current + spinOut * (direction || 1), {
                ...transition
            });
            return;
        }
        if (initialRef.current) {
            initialRef.current = false;
            if (!animateIn) return;
        }
        animate(current, cumulativeRef.current, {
            ...transition,
            delay
        });
    });
    return /*#__PURE__*/ jsxs("span", {
        style: {
            display: "inline-block",
            position: "relative",
            verticalAlign: "top"
        },
        children: [
            /*#__PURE__*/ jsx("span", {
                style: {
                    visibility: "hidden",
                    whiteSpace: "pre",
                    display: "inline-block"
                },
                children: "0"
            }),
            digits.map((n)=>/*#__PURE__*/ jsx(DigitNum, {
                    n: n,
                    current: current
                }, n))
        ]
    });
}
function SlotsRenderer({ text, Component, transition, stagger, animateInitial, className, style, rest }) {
    const chars = text.split("");
    const nextIdRef = useRef(chars.length);
    const [prevText, setPrevText] = useState(text);
    const [digitKeys, setDigitKeys] = useState(()=>chars.map((_, i)=>i));
    const dirRef = useRef(1);
    const mountedRef = useRef(false);
    useEffect(()=>{
        mountedRef.current = true;
    }, []);
    if (text !== prevText) {
        const result = reconcileDigitKeys(prevText, text, digitKeys, nextIdRef.current);
        nextIdRef.current = result.nextId;
        dirRef.current = result.direction;
        setDigitKeys(result.keys);
        setPrevText(text);
    }
    const dir = dirRef.current;
    const prefixLen = (()=>{
        const idx = chars.findIndex((c)=>isDigit(c));
        return idx === -1 ? chars.length : idx;
    })();
    const digitCount = chars.filter((c)=>isDigit(c)).length;
    let digitIndex = 0;
    return /*#__PURE__*/ jsx(MotionConfig, {
        transition: transition,
        children: /*#__PURE__*/ jsx(Component, {
            "aria-label": text,
            style: {
                display: "inline-flex",
                position: "relative",
                ...style
            },
            className: className,
            ...rest,
            children: /*#__PURE__*/ jsx("span", {
                style: {
                    display: "inline-flex",
                    paddingTop: FADE_HEIGHT,
                    paddingBottom: FADE_HEIGHT,
                    marginTop: `calc(-1 * ${FADE_HEIGHT})`,
                    marginBottom: `calc(-1 * ${FADE_HEIGHT})`,
                    maskImage: FADE_MASK,
                    WebkitMaskImage: FADE_MASK
                },
                children: /*#__PURE__*/ jsx(AnimatePresence, {
                    mode: "popLayout",
                    initial: animateInitial,
                    children: chars.map((char, i)=>{
                        const isPrefix = i < prefixLen;
                        const outerKey = isPrefix ? `pre-${i}` : `col-${chars.length - 1 - i}`;
                        if (isPrefix || !isDigit(char)) {
                            return /*#__PURE__*/ jsx(motion.span, {
                                layout: "position",
                                initial: false,
                                exit: isPrefix ? undefined : {
                                    opacity: 0
                                },
                                style: {
                                    display: "inline-block",
                                    whiteSpace: "pre"
                                },
                                children: char
                            }, outerKey);
                        }
                        const delay = (digitCount - 1 - digitIndex) * stagger;
                        digitIndex++;
                        return /*#__PURE__*/ jsx(motion.span, {
                            layout: "position",
                            initial: false,
                            exit: {
                                opacity: 0
                            },
                            style: {
                                display: "inline-block"
                            },
                            children: /*#__PURE__*/ jsx(SlotColumn, {
                                digit: Number(char),
                                direction: dir,
                                transition: transition,
                                delay: delay,
                                animateIn: mountedRef.current || animateInitial
                            })
                        }, outerKey);
                    })
                })
            })
        })
    });
}

function TextRenderer({ text, Component, transition, driftX, driftY, animateInitial, onComplete, className, style, rest }) {
    const nextIdRef = useRef(text.length);
    const [prevText, setPrevText] = useState(text);
    const [charKeys, setCharKeys] = useState(()=>text.split("").map((_, i)=>`c${i}`));
    const [changeRatio, setChangeRatio] = useState(0);
    if (text !== prevText) {
        const result = reconcileTextKeys(prevText, text, charKeys, nextIdRef.current);
        nextIdRef.current = result.nextId;
        setPrevText(text);
        setCharKeys(result.keys);
        setChangeRatio(result.changeRatio);
    }
    return /*#__PURE__*/ jsx(MotionConfig, {
        transition: transition,
        children: /*#__PURE__*/ jsx(Component, {
            "aria-label": text,
            style: {
                display: "inline-flex",
                ...style
            },
            className: className,
            ...rest,
            children: /*#__PURE__*/ jsx(AnimatePresence, {
                mode: "popLayout",
                initial: animateInitial,
                children: text.split("").map((char, i)=>{
                    const key = charKeys[i];
                    const progress = text.length <= 1 ? 0 : i / (text.length - 1);
                    const offsetX = (progress - 0.5) * driftX * changeRatio;
                    const offsetY = (progress - 0.5) * driftY * changeRatio;
                    const isLast = i === text.length - 1;
                    return /*#__PURE__*/ jsx(motion.span, {
                        "aria-hidden": "true",
                        layout: "position",
                        initial: {
                            opacity: 0,
                            x: offsetX,
                            y: offsetY
                        },
                        animate: {
                            opacity: 1,
                            x: 0,
                            y: 0
                        },
                        exit: {
                            opacity: 0,
                            x: offsetX,
                            y: offsetY
                        },
                        onAnimationComplete: isLast && onComplete ? onComplete : undefined,
                        style: {
                            display: "inline-block",
                            whiteSpace: "pre"
                        },
                        children: char
                    }, key);
                })
            })
        })
    });
}

function AutoSizeWrapper({ children, transition }) {
    const [element, setElement] = useState(null);
    const [width, setWidth] = useState(0);
    const ref = useCallback((node)=>{
        setElement(node);
    }, []);
    useEffect(()=>{
        if (!element) return;
        const observer = new ResizeObserver(([entry])=>{
            setWidth(entry.contentRect.width);
        });
        observer.observe(element);
        return ()=>observer.disconnect();
    }, [
        element
    ]);
    return /*#__PURE__*/ jsx(motion.span, {
        animate: {
            width: width > 0 ? width : "auto"
        },
        transition: transition,
        style: {
            display: "inline-flex",
            overflow: "hidden"
        },
        children: /*#__PURE__*/ jsx("span", {
            ref: ref,
            style: {
                display: "inline-flex"
            },
            children: children
        })
    });
}
/**
 * Calligraph — {@link https://calligraph.raphaelsalaja.com | Documentation}
 *
 * Fluid text and number transitions powered by Motion.
 *
 * @param props.variant - `"text"` for LCS character diffing, `"number"` for
 * rolling vertical digits, `"slots"` for slot-machine digit spin.
 * Defaults to `"text"`.
 *
 * @param props.animation - Spring preset: `"smooth"`, `"snappy"`, or
 * `"bouncy"`. Defaults to `"smooth"` for text, `"snappy"` for number.
 *
 * @param props.as - Wrapper element type. Defaults to `"span"`.
 *
 * @param props.drift - Maximum spread in pixels for entering/exiting
 * characters. `{ x, y }` — scaled by the fraction of characters that changed.
 * Only applies to `variant="text"`. Defaults to `{ x: 15, y: 0 }`.
 *
 * @param props.stagger - Seconds of delay spread across characters.
 * Defaults to `0.02`.
 *
 * @param props.initial - When `true`, characters animate in on first mount.
 * Defaults to `false`.
 *
 * @param props.onComplete - Fired when the last character finishes animating.
 *
 * @param props.autoSize - When `true`, the wrapper animates its width to
 * match content. Defaults to `true`.
 */ function Calligraph(props) {
    const { children, variant = "text", animation, as: Component = "span", drift: { x: driftX = 15, y: driftY = 0 } = {}, stagger = 0.02, initial: animateInitial = false, onComplete, autoSize = true, className, style, ...rest } = props;
    const transition = animations[animation ?? (variant === "number" ? "snappy" : "default")];
    const rendererProps = {
        text: String(children ?? ""),
        Component,
        transition,
        stagger,
        animateInitial,
        onComplete,
        className,
        style,
        rest
    };
    let content;
    if (variant === "number") {
        content = /*#__PURE__*/ jsx(NumberRenderer, {
            ...rendererProps
        });
    } else if (variant === "slots") {
        content = /*#__PURE__*/ jsx(SlotsRenderer, {
            ...rendererProps
        });
    } else {
        content = /*#__PURE__*/ jsx(TextRenderer, {
            ...rendererProps,
            driftX: driftX,
            driftY: driftY
        });
    }
    if (autoSize) {
        return /*#__PURE__*/ jsx(AutoSizeWrapper, {
            transition: transition,
            children: content
        });
    }
    return content;
}

export { Calligraph };
